rrdd/design/plugin-protocol-v3.md

```diff
@@ -0,0 +1,81 @@
+---
+title: RRDD plugin protocol v3
+layout: default
+design_doc: true
+revision: 1
+status: proposed
+revision_history:
+- revision_number: 1
+  description: Initial version
+---
+
+Motivation
+----------
+
+rrdd plugins protocol v2 report datasources via shared-memory file, however it
+has various limitations :
+ - metrics are unique by their names, thus it is not possible cannot have
```

Comment:
Don't the current data sources go around this by attaching the VM UUID to the name?

Reply:
I think the current protocol has these weaknesses in addition:
* not properly versioned; the header should have a fixed structure and include versioning information such that the protocol can be evolved
* The shared memory is allocated ahead of time, no matter how much space is actually needed, and can't be adjusted. It thus has to plan for the worst case.
* Combination of a binary format with an ASCII format - thus not really taking advantage of either: readability and cross platform vs. compactness.

One of the better ideas is probably that it relies on a checksum; the buffer can be updated concurrently but this is detected by the checksum. which could be written atomically. However, I am not sure the checksum is properly aligned to guarantee atomic update.

All in all, I think the current protocol's design is weak and I'm happy to see a proposal for something better.